scripts/west_commands/build.py

```diff
@@ -217,6 +217,18 @@ def do_run(self, args, remainder):
                         self.args.cmake_only or self.args.snippets):
                     self.run_cmake = True
         else:
+            if pristine == 'always':
+                # Look for a domains.yaml file in the build directory,
```

Comment:
Not sure which issue you try to fix here.
I can read following in commit message:
> This is because the top-level build directory was not being recognized as a Zephyr build directory, due to not having a `CMakeCache.txt` file yet.

The CMakeCache file is created immediately, so even if your CMakeLists.txt file only contains:
```
message(FATAL_ERROR "fail immediately")
```
then a CMakeCache.txt is created in the build folder.
Now, `west build --pristine` looks for the `ZEPHYR_BASE` setting in the CMake cache, which is set as part of `find_package(Zephyr)`.
Ref:
https://github.com/zephyrproject-rtos/zephyr/blob/54f848cfaa4ec5f46e61790f03b95c00ca107b71/scripts/west_commands/build_helpers.py#L123-L135

https://github.com/zephyrproject-rtos/zephyr/blob/54f848cfaa4ec5f46e61790f03b95c00ca107b71/share/sysbuild/CMakeLists.txt#L23

so if things are not working correctly then I would investigate why, instead of trying to add yet another file.

Especially a file which is created later in time than the `find_package(Zephyr)`

Reply:
> The CMakeCache file is created immediately

No, it's not created until CMake exits normally, i.e., configuration succeeds or fails. If you kill the process beforehand, then the file will not be there.

> Not sure which issue you try to fix here.

I'll explain it step by step. Suppose I start by running:
```
west build -p --sysbuild -b nrf9160dk_nrf9160 tests/boot/test_mcuboot
```
This will build three images: `test_mcuboot`, `mcuboot`, and `swapped_app`.

I let `test_mcuboot` and `mcuboot` finish generation, but I decide to hit Ctrl+C while generating `swapped_app`.

Now, I have two complete Zephyr build systems in `build/test_mcuboot` and `build/mcuboot`.

But, I want to build pristine again, so I rerun the same command. However, since the previous command was aborted, `build` is not recognized as a Zephyr build system, so it is not emptied.

`test_mcuboot` and `mcuboot` are also not emptied, so they are rebuilt from cache, which is not what I want. I consider this a failure to build pristine.

Reply:
Thanks for prividing more details on this matter.

> I let `test_mcuboot` and `mcuboot` finish generation, but I decide to hit Ctrl+C while generating `swapped_app`.
> 
> Now, I have two complete Zephyr build systems in `build/test_mcuboot` and `build/mcuboot`.
> 
> But, I want to build pristine again, so I rerun the same command. However, since the previous command was aborted, `build` is not recognized as a Zephyr build system, so it is not emptied.

The same happens on a regular Zephyr build (that is `--no-sysbuild`).
If you hit CTRL-C during CMake configure time, then the CMakeCache.txt is not created and a pristine is not carried out.
The reason why one may not immediately notice is because the CMakeCache.txt is also missing so no knowledge of an earlier run is present, but this doesn't mean everything will work as expected.
For example cases where a given file is only created if missing.
Such a left-over file from an aborted run can actually lead to an un-expected and hard to debug issue.

Your sysbuild use-case has put spot-light on this issue, and I acknowledge that the problem with extra images having a old CMakeCache.txt file present means there is a bigger risk someone might experience this problem.

But I would like a proper fix for the pristine feature instead of relying on a sysbuild specific file being present.

Reply:
> The same happens on a regular Zephyr build (that is `--no-sysbuild`).
> If you hit CTRL-C during CMake configure time, then the CMakeCache.txt is not created and a pristine is not carried out.
> The reason why one may not immediately notice is because the CMakeCache.txt is also missing so no knowledge of an earlier run is present, but this doesn't mean everything will work as expected.
> For example cases where a given file is only created if missing.
> Such a left-over file from an aborted run can actually lead to an un-expected and hard to debug issue.

Good point, I didn't think of that. I'm convinced now that this problem with pristine builds extends beyond sysbuild, so it should be attacked in a different way.

I will undo this work, leaving the domain schema changes. That's the more important stuff I wanted to get in for v3.5.0.

Reply:
This has been a long standing issue, there is even a thread on it somewhere, I don't remember if it's here or west or somewhere else but it described the issue and has a response from Marti